### DISCUSSION TITLE
automate asserting network payload data

### DISCUSSION THREAD

**Discussion by Kname80** [cypress-io/cypress#21051] [Questions and Help] [answered]
Hi Im new to cypress , I need to assert/validate the payload data , is it possible ? if yes can you please provide an example 

![image](https://user-images.githubusercontent.com/41926411/163014866-7a9cc647-9380-4c50-b28f-e627ce98425b.png)

**Comment by jjhelguero** [comment, marked as answer]
Yes, you can assert the [request object before it is sent](https://docs.cypress.io/api/commands/intercept#Request-Response-Modification-with-routeHandler). You'll use a callback function and make request body.

```js
// match requests method and url
cy.intercept('POST', '/users', (req) => {
  // make assertions on keys or contents
  expect(req.body).to.include.key('qryEmployeeForm')
  // make an assertion on the payload contents
  expect(req.body.qryEmployeeForm.orderBy).to.eq('NAME')
})
```

**Comment by Kname80** [comment]
wow this is awesome thank you for this much appreciated quick question how do you assert on something like this :   

{
"dogname :"time",
dogact :"some"
dogcolor :"something", 
  "food:{
      "foodone :"chicken",
      foodtwo:cheese,
      **"dogowner"  :["code 12332"],**
}
I would like to assert that "dogowner"  has the exact value of [
                   "code 12332"
                   ], 

  expect(req.body).to.include.key('["code 12332"]')
  // make an assertion on the payload contents
  expect(req.body.food.dogowner).to.eq('["code 12332"]')
Im getting this error 

expected  ["code 12332"] to contain key code 12332

**Reply by jjhelguero** [reply]
What you are doing with `expect(req.body.food.dogowner).to.eq('["code 12332"]')` is asserting it to have a string when it has array of strings. So, you'll want to assert it is an array first then drill down to the first element of the array and assert the string.

```js
expect(req.body.food.dogowner).to.be.instanceOf(Array)
expect(req.body.food.dogowner[0]).to.eq("code 12332")
```

**Comment by Kname80** [comment]
for some reason instance of (Array) didn't work, but this work form me ,
expect(req.body.food.dogowner[0]) .to.be.a('array')
thank you for your help , I'm new to Cypress still learning

**Reply by Kname80** [reply]
also using deep as stated here https://github.com/chaijs/chai/issues/1062 solved it

**Reply by Kname80** [reply]
expect(req.body).to.include.key('["code 12332"]')
// make an assertion on the payload contents
expect(req.body.food.dogowner).to.eq('["code 12332"]')

**Reply by Kname80** [reply]
sorry one more thing, 
what about if we have  the same thing but with an array that includes 200 or more items that dynamically changes

{
"dogname :"time",
dogact :"some"
dogcolor :"something",
"food:{
"foodone :"chicken",
foodtwo:cheese,
**"dogowner" :["code 12332" , "code 5445", " code 4555", "code 5554 .......................................],** 
every time you check the request the code value array change and more codes are added to the payload array?
}
how do we assert on something like this ?

**Reply by jjhelguero** [reply]
You can use regex to assert each element in the array matches 'code ' followed by 5 digits, `/code\s\d{5}/`. Of the top of my head, you can use an array for each loop to check them, but there might be a better way to do it.

**Reply by Kname80** [reply]
Nice , great idea I will try it out
thanks
